automated/linux/mmtests/mmtests.sh

```diff
@@ -0,0 +1,151 @@
+
+	<TEST_GIT_URL>:
+	If this parameter is set, then the ${TEST_PROGRAM} suite is cloned
+	from the URL in TEST_GIT_URL. Otherwise it is cloned from the
+	standard repository for the suite. Note that cloning is done
+	only if TEST_PROG_VERSION is not empty
+
+	<TEST_DIR>:
+	If this parameter is set, then the ${TEST_PROGRAM} suite is cloned to or
+	looked for in TEST_DIR. Otherwise it is cloned to $(pwd)/${TEST_PROGRAM}
+
+	<SKIP_INSTALL>:
+	If you already have it installed into the rootfs.
+	default: false
+
+	<MMTESTS_CONFIG_FILE>:
+	Mmtests configuration file that describes how the benchmarks should be
+	configured and executed.
+
+	<MMTESTS_TYPE_NAME>:
+	Mtests test type, e.g. sysbenchcpu, iozone, sqlite, etc.
+
+	<MMTESTS_MAX_RETRIES>:
+	Maximum number of retries for the single benchamrk's source file download"
+	exit 1
+}
+
+while getopts "c:p:r:s:t:u:v:" opt; do
+	case "${opt}" in
+		c)
+			MMTESTS_CONFIG_FILE="${OPTARG}"
+			;;
+		p)
+			if [[ "$OPTARG" != '' ]]; then
+				TEST_DIR="$OPTARG"
+			fi
+			;;
+		r)
+			MMTESTS_MAX_RETRIES="${OPTARG}"
+			;;
+		s)
+			SKIP_INSTALL="${OPTARG}"
+			;;
+		t)
+			MMTESTS_TYPE_NAME="${OPTARG}"
+			;;
+		u)
+			if [[ "$OPTARG" != '' ]]; then
+				TEST_GIT_URL="$OPTARG"
+			fi
+			;;
+		v)
+			TEST_PROG_VERSION="$OPTARG"
+			;;
+		*)
+			usage
+			;;
+	esac
+done
+
+install() {
+	dist=
+	dist_name
+	case "${dist}" in
+	debian|ubuntu)
+			pkgs="build-essential wget perl git autoconf automake \
+					bc binutils-dev btrfs-progs linux-cpupower expect \
+					gcc hdparm hwloc-nox libtool numactl tcl time \
+					xfsprogs xfslibs-dev libopenmpi-dev"
+			install_deps "${pkgs}" "${SKIP_INSTALL}"
+		;;
+	fedora|centos)
+		pkgs="git gcc make automake libtool wget perl autoconf \
+					bc binutils-devel btrfs-progs kernel-tools expect \
+					hdparm hwloc libtool numactl tcl time xfsprogs \
+					openmpi-devel"
+			install_deps "${pkgs}" "${SKIP_INSTALL}"
+		;;
+	oe-rpb)
+		# Assume all dependent packages are already installed.
+		;;
+	*)
+		warn_msg "Unsupported distro: ${dist}! Package installation skipped!"
+		;;
+	esac
+	PERL_MM_USE_DEFAULT=1
+	export PERL_MM_USE_DEFAULT
+	cpan -F JSON Cpanel::JSON::XS List::BinarySearch
+	AUTO_PACKAGE_INSTALL=yes
+	export AUTO_PACKAGE_INSTALL
+	DOWNLOADED=0
+	COUNTER=0
+	while [ $DOWNLOADED -eq 0 -a $COUNTER -lt $MMTESTS_MAX_RETRIES ]; do
+		./run-mmtests.sh -b --no-monitor --config ${MMTESTS_CONFIG_FILE} benchmark && DOWNLOADED=1
+		COUNTER=$((COUNTER+1))
+	done
+}
+
+run_test() {
+	info_msg "Running ${MMTESTS_TYPE_NAME} cpu test..."
+	extracted_json="../${MMTESTS_TYPE_NAME}.json"
+	./run-mmtests.sh --no-monitor --config ${MMTESTS_CONFIG_FILE} benchmark
+	./bin/extract-mmtests.pl -d work/log/ -b ${MMTESTS_TYPE_NAME} -n benchmark --print-json > $extracted_json
+}
+
+! check_root && error_msg "Please run this script as root."
+
+get_test_program "${TEST_GIT_URL}" "${TEST_DIR}" "${TEST_PROG_VERSION}" "${TEST_PROGRAM}"
+
+# Test installation.
```

Comment:
This if-else block must be before the function 'get_test_program()'

Reply:
Doing that would break the while loop at line 126 which is needed to download correctly the test source files (sometimes the mirror replies with a 404 code instead of sending the file) and, in a way, is part of the installation process.
I can put that part of code in a different function that can be called later, is that ok?